# Modele językowe - n-gramy 

---

## 1. N-gramy słów w klasyfikacji
Poniżej stworzono kod, który przeprowadza klasyfikację dokumentów należących do 4 kategorii. W odróżnieniu do poprzednich zajęć - tu zaproponowano klasyfikator SVC (algorytm SVM, popularna alternatywa dla NaiveBayes), która również świetnie się spisuje w problemach klasyfikacji tekstu.

**<span style="color: red">Zadanie 1a (0.5 punktu)</span>** Uruchom kod, przyjrzyj się wygenerowanym wynikom (a najlepiej zachowaj je gdzieś, będą potrzebne). <br/>
Zapoznaj się z dokumentacją TfIdfVectorizer, odnajdź opcję uwzględnienia nie tylko pojedynczych słów jako cechy, ale także ich par i **zmodyfikuj poniższy kod tak, aby klasyfikacja uwzględniała zarówno pojedyncze słowa jak i pary (pozostaw parametr max_df=0.1 nienaruszony).** <span style="color: red"> Zmodyfikuj linię 30.</span><br/> <br/>
**<span style="color: red">Zadanie 1b (0.5 punktu)</span>** Jak zmieniła się liczba cech po uwzględnieniu tych par? Czy coś się zmieniło w raporcie z klasyfikacji? Uzupełnij odpowiedzi na pytania w komórce poniżej kodu.

In [2]:
from sklearn.pipeline import Pipeline
from sklearn.svm import SVC
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics import classification_report
from sklearn.datasets import fetch_20newsgroups # zbiór danych zawarty w Sklearn, który zawiera dane z 20 grup newsowych
import numpy as np

# ------------------- WCZYTANIE DANYCH -----------

np.random.seed(0) # ustaw seed na 0, aby zapewnić powtarzalność eksperymentu

categories = ['sci.space', 'comp.graphics', 'talk.politics.misc', 'comp.sys.mac.hardware'] # kategorie do analizy

train = fetch_20newsgroups(subset='train',
                                   categories=categories,
                                   shuffle=True,
                                   random_state=42) # pobieramy zbiór uczący (na nim będziemy trenować) dla wybranych kategorii.
    

test = fetch_20newsgroups(subset='test',
                                  categories=categories,
                                  shuffle=True,
                                  random_state=42) # pobieramy zbiór testowy (na nim będziemy testować) dla wybranych kategorii



# ------------------- STWORZENIE PIPELINE'U -----------
    
pipeline = Pipeline([             # stwórzmy pipeline surowy tekst -> TFIDF vectorizer -> klasyfikator 
    ('tfidf', TfidfVectorizer(max_df=0.1, ngram_range=(1,2))), # linijka 30
    ('clf', SVC(C=1.0, kernel='linear')),
])

# ------------------- TRANSFORMACJA I UCZENIE -----------

pipeline.fit(train.data, train.target) # zwektoryzujmy dane i wytrenujmy klasyfikator na zbiorze treningowym


print("W słowniku znajduje się {n} różnych cech".format(
    n=len(pipeline.named_steps['tfidf'].vocabulary_.keys())
))

# ------------------- OCENA KLASYFIKATORA -----------
print(classification_report(test.target, pipeline.predict(test.data))) # testowanie klasyfikatora - szerokie podsumowanie uwzględniające miary: precision, recall, f1

W słowniku znajduje się 287718 różnych cech
              precision    recall  f1-score   support

           0       0.87      0.94      0.91       389
           1       0.92      0.94      0.93       385
           2       0.95      0.90      0.92       394
           3       0.97      0.90      0.93       310

    accuracy                           0.92      1478
   macro avg       0.93      0.92      0.92      1478
weighted avg       0.92      0.92      0.92      1478



Zapamiętane wyniki:

W słowniku znajduje się 34774 różnych cech
              precision    recall  f1-score   support

           0       0.88      0.93      0.90       389
           1       0.91      0.93      0.92       385
           2       0.94      0.91      0.92       394
           3       0.96      0.90      0.93       310

    accuracy                           0.92      1478
   macro avg       0.92      0.92      0.92      1478
   
weighted avg       0.92      0.92      0.92      1478

<span style="color: red"> 1. O ile zwiększyła się liczba cech w klasyfikatorze? </span> Znacząco wzrosła z 34774 do 287718. <br> 
<span style="color: red"> 2. Czy precyzja w którejkolwiek klasie wzrosła? w której/których? </span> Tak, w 1 i 3. <br>
<span style="color: red"> 3. Czy recall w którejkolwiek klasie wzrósł? w której/których? </span> Tak, w 0 i 1. <br>


## 2. N-gramy liter w klasyfikacji
Poza n-gramami stworzonymi z następujących po sobie wyrazów - bardzo często używane są również n-gramy znakowe, stworzone z następujących po sobie liter. <br/><br/>
Dla przykładu. wszystkie 3-gramy (trigramy) znakowe z napisu "Hello world" to: <br/>
"Hel", "ell", "llo", "lo ", "o w", " wo", "wor", "orl", "rld". <br/><br/>
Do czego mogłaby być użyta taka reprezentacja tekstów? Okazuje się, że całkiem mocno pomaga to w rozwiązywaniiu problemu detekcji języka w którym został zapisany dokument, szczególnie w sytuacji, kiedy teksty są bardzo krótkie (np. tweety, smsy).
<br/>
Poniżej znajduje się szkielet klasyfikatora rozpoznającego język w którym zapisany jest dokument.
Języków jest 6: polski, angielski, niemiecki, francuski, hiszpański i włoski.
<br/>
**<span style="color: red">Zadanie 2 (1 punkt)</span>**: Przedstawiony klasyfikator jest znanym już z poprzednich przykładów kodem. Waszym zadaniem jest:
<ol>
    <li>Zapoznanie się dokumentacją Tf-Idf vectorizera, aby znaleźć funkcjonalność, która zamiast całych słów, stworzy cechy na podstawie liter i wykorzystanie tej funkcjonalności w kodzie</li>
    <li>Ustawienie takiego zakresu n-gramów, aby zmaksymalizować uzyskany wynik (Oczekiwane 1.0 precyzji i recallu we wszystkich kategoriach przy pozostawieniu wartośi max_features = 300 elementów)</li>
    <li>Poprawnie zaklasyfikuje krotki przykład zapisany w linii 43 (Bonjour przypisze do kategorii 'french').</li>
</ol>

In [3]:
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report
import pandas
import numpy as np

# -------------------- FUNKCJE POMOCNICZNE --------

# Funkcja mapująca identyfikator liczbowy kategorii na wartość tekstową, np: 0->"polish", 1->"english"
def get_class_name_from_id(ids, mapping):
    return [mapping[id] for id in ids]
# ------------------- WCZYTANIE DANYCH -----------
full_dataset = pandas.read_csv('language_detection_1000.csv', encoding='utf-8') # wczytaj dane z pliku CSV
lang_to_id = {'polish': 0, 'english': 1, 'french': 2,
              'german': 3, 'italian': 4, 'spanish': 5}
id_to_lang = {v: k for k,v in lang_to_id.items()}
full_dataset['label_num'] = full_dataset.lang.map(lang_to_id)  # ponieważ nazwy kategorii zapisane są z użyciem stringów: "ham"/"spam", wykonujemy mapowanie tych wartości na liczby, aby móc wykonać klasyfikację. 

np.random.seed(0)                                       # ustaw seed na 0, aby zapewnić powtarzalność eksperymentu
train_indices = np.random.rand(len(full_dataset)) < 0.7 # wylosuj 70% wierszy, które znajdą się w zbiorze treningowym

train = full_dataset[train_indices] # wybierz zbior treningowy (70%)
test = full_dataset[~train_indices] # wybierz zbiór testowy (dopełnienie treningowego - 30%)


# ------------------- STWORZENIE PIPELINE'U -----------  
pipeline = Pipeline([             # stwórzmy pipeline surowy tekst -> TFIDF vectorizer -> klasyfikator 
    ('tfidf', TfidfVectorizer(max_features=300, ngram_range=(3,3), analyzer='char_wb')),
    ('scaler', StandardScaler(with_mean = False)),
    ('clf', LogisticRegression()),
])
# ------------------- TRANSFORMACJA I UCZENIE -----------

pipeline.fit(train['text'], train['label_num']) # zwektoryzujmy dane i wytrenujmy klasyfikator na zbiorze treningowym

print("Oto kilka przykładowych cech stworzonych przez TfidfVectorizer: {n}".format(
    n=list(pipeline.named_steps['tfidf'].vocabulary_.keys())[:5]))

# ------------------- WERYFIKACJA NA KRÓTKIM TEKŚCIE ----

text_to_predict = "Bonjour!"
predicted = pipeline.predict([text_to_predict])
print("\n\nTekst: {t} został zaklasyfikowany jako: {p}\n\n".format(
    t=text_to_predict,
    p=id_to_lang[predicted[0]]
))


# ------------------- OCENA KLASYFIKATORA -----------
print(classification_report(
    get_class_name_from_id(test['label_num'], id_to_lang), 
    get_class_name_from_id(pipeline.predict(test['text']), id_to_lang)
))

Oto kilka przykładowych cech stworzonych przez TfidfVectorizer: [' do', 'do ', ' wi', 'prz', 'rze']


Tekst: Bonjour! został zaklasyfikowany jako: french


              precision    recall  f1-score   support

     english       1.00      1.00      1.00       303
      french       1.00      1.00      1.00       280
      german       1.00      1.00      1.00       337
     italian       1.00      1.00      1.00       273
      polish       1.00      1.00      1.00       291
     spanish       1.00      1.00      1.00       299

    accuracy                           1.00      1783
   macro avg       1.00      1.00      1.00      1783
weighted avg       1.00      1.00      1.00      1783



Widzimy, że problem jest stosunkowo prosty. Po co zatem używać n-gramów znakowych? Aby zaoszczędzić pamięć i podołać sytuacjom, w których zbiór testowy składa się ze słów, które nie występują w korpusie uczącym. <br/>

O ile wszystkich słów w tych 6 językach jest "30078", to trigramów znakowych jest już tylko "15274", a bigramów - "2059". W związku z tym: <ol>
<li>Używając n-gramów znakowych często możemy ograniczyć liczbę cech</li>
<li>N-gramy znakowe pomogą nam w sytuacjach, kiedy dane słowo nie wystąpiło w tekście uczącym. Jeśli opieramy uczenie na pełnych słowach i cały nasz tekst testowy to niewystępujące w korpusie uczącym - "bonjour", wtedy wektor BOW będzie zawierał same zera, przez co będzie miał problem z przydziałem do odpowiedniej klasy. <br/> N-gramy znakowe nawet jeśli nie napotkały danego słowa podczas analizy korpusu, to na podstawie budowy samego słowa są w stanie przewidywać do jakiego języka słowo należy. Np. cokolwiek zawierającego trigram "żeb" należeć będzie raczej do języka polskiego.</li>
</ol>

---

### 2a - istotność cech

Ponieważ w zadaniu 2 użyliśmy znanego z zajęć z przedmiotu "Sztuczna Inteligencja" klasyfikatora liniowego - regresji logistycznej, podejrzeć możemy jakie cechy najsilniej sugerują nam przynależność do danej klasy. Uruchom poniższy kod, aby zobaczyć jakie cechy są najważniejsze dla danych kategorii. Modyfikując parametry TfidfVectorizer możesz zobaczyć jakie słowa/ciągi znaków są najistotniejsze do detekcji danego języka.

In [5]:
# Funkcja, z użyciem której możemy ocenić które cechy najsilniej skojarzone są z danymi klasami.
# Wyświetli listę słów/n-gramów znakowych, których obecność najsilniej wpływa na przydział do danej klasy
def language_indicators(feature_names, feature_importances, id_to_lang):
    for i, language in enumerate(feature_importances): # iterujemy po macierzy feature_importances (wymiarów: język x cechy) wierszami (czyli język po języku)
        scored_features = list(zip(feature_names, language)) # tworzymy skojarzenie nazw cech z wagami modelu (ponieważ używamy regresji logistycznej - każda cecha (słowo/n-gram) ma swoją wagę, która jest optymalizowana w procesie uczenia! Cechy z wysokimi wagami są ważne dla danej klasy. Każda klasa ma osobny model ze swoimi wagami!)
        scored_features = sorted(scored_features, key=lambda x: x[1], reverse=True) # posortujmy cechy skojarzone z wagami malejąco 
        print("W rozpoznaniu języka {lang} najważniejsze cechy to:".format(
            lang=id_to_lang[i]) #zamieńmy identyfikator numeryczny kategorii na nazwę języka
        )
        for feature, score in scored_features[:5]: # wybierzmy po 5 najważniejszych cech (najwyższe wartości uczonych współczynników)
            print("\t'{feature}': {score}".format(feature=feature, score=score))
        

# ------------------- WYŚWIETLENIE NAJWAŻNIEJSZYCH CECH DLA KAŻDEJ KATEGORII
language_indicators(
    pipeline.named_steps['tfidf'].get_feature_names_out(), # pobierz nazwy cech
    pipeline.named_steps['clf'].coef_, # pobierz wyuczone współczynniki (regresja logistyczna to stworzenie modelu opisanego wzorem y = e^(-wx - b), gdzie uczymy się współczynników w. Pole coef_ zawiera te współczynniki dla każdego języka z osobna)
    id_to_lang # mapowanie z identyfikatora numerycznego na pełną nazwę języka - zwiększa czytelność wygenerowanego raportu
)

W rozpoznaniu języka polish najważniejsze cechy to:
	' w ': 0.2897523578307207
	'dzi': 0.2623696727452865
	' po': 0.24653825702443577
	'owa': 0.24349982093287356
	'rze': 0.2338494676925721
W rozpoznaniu języka english najważniejsze cechy to:
	'the': 0.4408454804664215
	' th': 0.41405398063555277
	'ing': 0.271858271418333
	' an': 0.26851892824620105
	'ed ': 0.2543240806703816
W rozpoznaniu języka french najważniejsze cechy to:
	' ce': 0.3167282684769287
	' à ': 0.26087793588261954
	'ce ': 0.260265426215297
	' et': 0.24470420947649163
	'nt ': 0.23515278929622102
W rozpoznaniu języka german najważniejsze cechy to:
	'der': 0.3252823040306388
	'die': 0.25592666806156383
	'en ': 0.25150810316920186
	' ge': 0.22961029312957934
	'er ': 0.2288158660596224
W rozpoznaniu języka italian najważniejsze cechy to:
	'di ': 0.39076479901219885
	' e ': 0.3143000441042705
	'lla': 0.25708490932132716
	'zio': 0.24676876532474845
	'ti ': 0.22825473272397032
W rozpoznaniu języka spanish najważniejsze cechy to

---

## 3. N-gramy słów w generowaniu tekstu

Innym, bardzo ciekawym zastosowaniem n-gramów jest możliwość generowania tekstu z użyciem tzw. łańcuchów Markova. Stwórzmy funkcję generującą n-gramy słów, aby później móc ją wykorzystać do tworzenia tekstów.

**<span style="color: red">Zadanie 3 (1 punkt)</span>** stwórz funkcję, która wygeneruje n-gramy słów zadanego stopnia n (n_gram_len). Aby podzielić zdanie na słowa nie musisz używać tokenizatora z biblioteki, na potrzeby zadania wystarczy uznać, że spacja oddziela poszczególne słowa.
<br/>
<br/>
<div class="alert alert-success">
Oczekiwany rezultat dla zadanych danych: <br/><br/>[['The', 'big', 'brown'], ['big', 'brown', 'fox'], ['brown', 'fox', 'jumped'], ['fox', 'jumped', 'over'], ['jumped', 'over', 'the'], ['over', 'the', 'fence.']]
</div>
<br/>

In [7]:
def get_word_ngrams(data, n_gram_len):
    words = data.split(' ')
    ngrams = []
    for i in range (len(words) - n_gram_len + 1):
        ngram = []
        for j in range(n_gram_len):
            ngram.append(words[i+j])
        ngrams.append(ngram)
    return ngrams
print(get_word_ngrams("The big brown fox jumped over the fence.", 3))

[['The', 'big', 'brown'], ['big', 'brown', 'fox'], ['brown', 'fox', 'jumped'], ['fox', 'jumped', 'over'], ['jumped', 'over', 'the'], ['over', 'the', 'fence.']]


Jeśli udało Ci się napisać funkcję get_word_ngrams - zapoznaj się z poniższym kodem i uruchom go, aby wytworzyć tekst!

In [9]:
from collections import Counter
import random
import itertools

def generate_ngram_markov(n_gram_len):
    markov_dict = dict() # stwórz słownik, który wskaże listę dozwolonych słów po zaobserwowanych n-poprzednich słowach.
    with open("polish_europarl.txt", 'r', encoding="utf8") as f: # wczytaj korpus danych
        data = f.read().lower()                                           # zamień wszystkie wielkie litery na małe
        n_grams = get_word_ngrams(data, n_gram_len)                   # wygeneruj wszystkie n-gramy słów z korpusu
        for n_gram in n_grams:                   # dla każdego n-gramu...
            context = " ".join(n_gram[:-1])      # weź wszystkie słowa z n-gramu poza ostatnim i połącz w 1 string spacją
            last_word = str(n_gram[-1])          # weź ostatnie słowo n-gramu
            
            if context not in markov_dict.keys(): # jeśli n-gram ubiedzony o ostatnie słowo nie występuje jeszcze w słowniku
                markov_dict[context] = list()     # to dopiszmy go do słownika i stwórzmy mu listę
            markov_dict[context].append(last_word) # wiedząc, że ubiedzony n-gram jest w słowniku - dopiszmy ostatnie słowo do listy
    
    for context in markov_dict.keys():                        # dla każdego kontekstu (ubiedzonego n-gramu)
        markov_dict[context] = Counter(markov_dict[context])  # stwórz histogram słów jakie występują w korpusie po tym kontekście
    
    return markov_dict


n_gram_len = 3  # liczba słów do stworznia n-gramu
markov_dict = generate_ngram_markov(n_gram_len)  # stworzenie słownika z histogramami słów dla poszczególnych kontekstów

text = 'Średnio co dwa' # tekst, od którego zaczniemy generowanie

for i in range(500):               # powtórzmy 500 razy ...
    text_spl = text.split(" ")     # podzielmy istniejący tekst po spacji (przeprowadźmy naiwną 'tokenizację')
    context = " ".join(text_spl[-n_gram_len+1:])   # pobierzmy ostatnie n_gram_len - 1 słów
    idx = random.randrange(sum(markov_dict[context].values())) # sprawdźmy słowa, które są dozwolone jako następniki naszego kontekstu (context) i wybierzmy taki następnik, który zostanie wylosowany zgodnie z rozkładem stworzonym przez histogram.
    new_word = next(itertools.islice(markov_dict[context].elements(), idx, None)) # wybierzmy wylosowane słowo
    text = text + " " + new_word # doklejmy wylosowane słowo na końcu
print(text)

Średnio co dwa miesiące od początku sprzeciwiali się utworzeniu esdz, uważając, że globalne partnerstwa mają charakter komercyjny, a zatem mniejszej biurokracji, a także nadmierna biurokracja w kraju.
skuteczne zarządzanie imigracją i korupcją, a osobiście nie wierzę, aby kontrole graniczne i, jeśli to prawda, że indie będą się utrzymywały, a taka jest ostatecznie reakcja europy.
pakt na rzecz stworzenia równego pola gry.
takie zbilansowanie płatności będzie nie tylko wymiar miejski nie zyskał jak dotąd w dużym stopniu zazębiają się z sześciu sytuacji, w której wielka brytania nie wprowadziła jednej waluty; ze względu na to, że firmy ponoszą społeczną odpowiedzialność wobec swoich pracowników do zatrudnienia i spraw socjalnych - panie przewodniczący! cieszę się, że do pewnego stopnia usankcjonowane w porozumieniach międzynarodowych i przedostanie się na wzajemnym zaufaniu, i wszystkie regiony europejskiej polityki handlu w wymiarze humanitarnym i problemem imigracji i proponować go inn

---

## 4. N-gramy znakowe w generowaniu tekstu

W bardzo podobny sposób do zadania 3, możemy stworzyć model, który generować będzie tekst literka po literce. <br/>
**<span style="color: red">Zadanie 4 (1 punkt)</span>** stwórz funkcję, która wygeneruje n-gramy znakowe zadanego stopnia n (n_gram_len).
<br/>
<br/>
<div class="alert alert-success">
Oczekiwany rezultat dla zadanych danych: ['The', 'he ', 'e b', ' bi', 'big', 'ig ', 'g b', ' br', 'bro', 'row', 'own', 'wn ', 'n f', ' fo', 'fox', 'ox ', 'x j', ' ju', 'jum', 'ump', 'mpe', 'ped', 'ed ', 'd o', ' ov', 'ove', 'ver', 'er ', 'r t', ' th', 'the', 'he ', 'e f', ' fe', 'fen', 'enc', 'nce', 'ce.']
</div>
<br/>

In [11]:
def get_character_ngrams(data, n_gram_len):
    ngrams = []
    for i in range(len(data) - n_gram_len + 1):
        ngrams.append(data[i:i+n_gram_len])
    return ngrams
print(get_character_ngrams("The big brown fox jumped over the fence.", 3))

['The', 'he ', 'e b', ' bi', 'big', 'ig ', 'g b', ' br', 'bro', 'row', 'own', 'wn ', 'n f', ' fo', 'fox', 'ox ', 'x j', ' ju', 'jum', 'ump', 'mpe', 'ped', 'ed ', 'd o', ' ov', 'ove', 'ver', 'er ', 'r t', ' th', 'the', 'he ', 'e f', ' fe', 'fen', 'enc', 'nce', 'ce.']


Po stworzeniu funkcji **get_character_ngrams()** możemy uruchomić generator znakowy.

In [13]:
from collections import Counter
import random
import itertools

def generate_ngram_markov(n_gram_len):
    markov_dict = dict() # stwórz słownik, który wskaże listę dozwolonych słów po zaobserwowanych n-poprzednich słowach.
    with open('pan_tadeusz.txt', 'r', encoding="utf8") as f: # wczytaj korpus danych
        data = f.read().lower()                                           # zamień wszystkie wielkie litery na małe
        n_grams = get_character_ngrams(data, n_gram_len)                   # wygeneruj wszystkie n-gramy słów z korpusu
        for n_gram in n_grams:                   # dla każdego n-gramu...
            context = n_gram[:-1]  # weź wszystkie znaki n-gramu poza ostatnim 
            last_char = n_gram[-1] # weź ostatni znak n-gramu
            if context not in markov_dict.keys(): # jeśli n-gram ubiedzony o ostatni znak nie występuje jeszcze w słowniku
                markov_dict[context] = list()     # to dopiszmy go do słownika i stwórzmy mu listę
            markov_dict[context].append(last_char) # wiedząc, że ubiedzony n-gram jest w słowniku - dopiszmy ostatni znak do listy
    
    for context in markov_dict.keys():                        # dla każdego kontekstu (ubiedzonego n-gramu)
        markov_dict[context] = Counter(markov_dict[context])  # stwórz histogram liter jakie występują w korpusie po tym kontekście
    
    return markov_dict


text = 'U szlachty' # tekst, od którego zaczniemy generowanie
n_gram_len = len(text)  # liczba znaków do stworznia n-gramu
markov_dict = generate_ngram_markov(n_gram_len)  # stworzenie słownika z histogramami słów dla poszczególnych kontekstów

for i in range(500):               # powtórzmy 500 razy ...
    context = text[-n_gram_len+1:]   # pobierzmy ostatnie n_gram_len - 1 słów
    idx = random.randrange(sum(markov_dict[context].values())) # sprawdźmy słowa, które są dozwolone jako następniki naszego kontekstu (context) i wybierzmy taki następnik, który zostanie wylosowany zgodnie z rozkładem stworzonym przez histogram.
    new_char = next(itertools.islice(markov_dict[context].elements(), idx, None)) # wybierzmy wylosowane słowo
    text = text + new_char # doklejmy wylosowany znak na końcu

print(text)

U szlachty tłuszcza;
bo pan bóg, kiedy karę na naród przypuszcza,
słania się pokornie,
dama raczyła się obrócić; dziewczyna wstydliwa
obraca się, lecz mu nie brak męstwa,
na miejscu swym siadła
pomiędzy piołunu i chwastu,
wznoszą się z nim stało, różnie powiadał, że sobie poradzić,
umyśliłem zakopać go razem z zamkiem oddać raczy».

    «jużci — bąknął tadeusz szczerze chciał zjeść izbę całą;
łatwo zgadnąć, że szlachcica.
czasem telimeny, czoło pochylony,
z palcami zadzwoniło w strunę skrzydło muchy,
wyda


---

## 5. Ngramy do generowania tekstu - długość ngramu a jakość tekstu
<span style="color: red">**Zadanie 5 (1 punkt)**</span>
Obswerując wyniki z zadań 3 i 4 i sprawdzając różne długości n-gramów (znakowych i słów) zastanów się:
<ol>
<li>Jakie ryzyko w kontekście jakości tekstu niesie ze sobą tworzenie tekstu z bardzo **krótkich** n-gramów?</li>
<li>Jakie ryzyko w kontekście jakości tekstu niesie ze sobą tworzenie tekstu z bardzo **długich** n-gramów?</li>
</ol>
Odpowiedzi zawrzyj w komentarzu poniżej

In [ ]:
# Zad 5:
#  Pytanie 1: Często generowany tekst nie ma sensu, bo brakuje informacji o kontekście i otoczeniu
#  Pytanie 2: Może pojawić się Key Error, gdy dana kombinacja nie występuje w słowniku

---

## 6. Bonus: Prawdopodobieństwo wystąpienia zdania - bez punktów
Dodatkowo ciekawym zastosowaniem n-gramów jest również ocena - jak bardzo prawdopodobnym jest wystąpienie danego zdania w rzeczywistości. Kiedy rozwiązujemy zadanie translacji mowy na tekst, spotykamy się z sytuacjami, w których nie do końca wiemy, czy słowo, które zostało wypowiedziane to np. "morze" czy "może". Model językowy oparty o n-gramy może ocenić szansę wystąpienia danego ciągu wyrazów, a więc również wybrać bardziej prawdopodobny ciąg wyrazów w danym języku. <br/>

Biorąc pod uwagę, że zdanie to ciąg wyrazów    $w_1, w_2, w_3, ..., w_n$
Możemy poczynić upraszczające założenie, że aktualne słowo zależne jest jedynie od słowa poprzedniego, zatem prawdopodobieństwo wystąpienia zdania $P(sentence) = p(w_1|beginOfSentence)*p(w_2|w_1)*p(w_3|w_2)*...*p(w_n|w_(n-1))$

Obliczając prawdopodobieństwa warunkowe, może się okazać, że w testowanym przez nas zdaniu mogą wystąpić dwie problematyczne sytuacje:
<ol>
    <li>słowo konteksowe $w_c$ ze wzoru $p(w_n|w_c)$ nie występuje w korpusie - bardzo mała szansa jeśli korpus jest wystarczająco duży</li>
    <li>słowo następujące po kontekstowym ($w_n$) nie współwystępuje w korpusie ze słowem kontekstowym (więc $p(w_n|w_c) = 0$ - całkiem możliwy stan, dość łatwo można sobie wyobrazić sytuację braku współwystępowania pewnych słów nawet przetwarzając duży korpus</li>
</ol>

Aby poradzić sobie z sytuacją, w której chcemy aby pewne słowo rozpoczynało/kończyło tekst, możemy dodać sztuczne znaczniki początku (BOS - Begin of Sentence) i końca (EOS - End of Sentence) zdania. Wprowadzając te znaczniki, będziemy mogli obliczyć prawdopodobieństwo wystąpienia słowa, pod warunkiem, że rozpoczyna ono zdanie $p(w_n|BOS)$

Poniżej znajduje się kod oceniający prawdopodobieństwo wystąpienia zdań. Widzimy, że jedno z tych zdań ma sensowniejszy tekst i chcielibyśmy, aby komputer był w stanie wybrać sensowniejszą opcję.

Problematyczne sytuacje rozwiązane zostały następująco:
<ol>
<li>Jeśli brak słowa kontekstowego w wygenerowanym modelu - uznaj, że prawdopodobieństwo całego zania wynosi 0</li>
<li>Jeśli słowo następujące po kontekstowym nie współwystępuje z kontekstowym - użyj wygładzania aby ustawić prawdopodobieństwo na bardzo małą wartość (ale niezerową)</li>
</ol>

**Zapoznaj się z kodem i urochom go, tutaj nie trzeba nic zmieniać, to jedynie wizualizacja zastosowania. Uwaga - aby poprawnie oszacować prawdopodobieństwa potrzeba posiadać wykonane zadanie 3 (stworzona funkcja get_word_ngrams)**


In [15]:
text1 = "i heard that the european union is a valuable concept." # tekst do oceny
text2 = "i had that the euro bean union is a variable concept."  # tekst do oceny

from nltk import sent_tokenize                                   # będziemy dzielić na zdania
import re                                                        # i czyścić tekst

from collections import Counter, defaultdict
import random
import itertools

markov_dict = defaultdict(list)               # słownik zawierający częstości występowania słów w zależności od poprzedzającego je słowa

def clean_text(text):
    return re.sub("[\n\t ]+", " ", text) # czyszczenie tekstu ze znaków nowej linii, tabulatorów, spacji (wielokrotnych)

def make_begin_end_of_sentences(text):
    result = ""
    sentences = sent_tokenize(text)
    for sent in sentences:
        result += " <BOS> {s} <EOS> ".format(s=sent) # dla każdego zdania dodajemy specjalne tagi <BOS> = begin of sentence oraz <EOS> - end of sentence
    return clean_text(result)

def get_sentence_probability(sentence, markov_dict):
    sentence = " <BOS> {s} <EOS> ".format(s=sentence)
    sentence = clean_text(sentence)
    
    sentence = sentence.split(' ')
    prob = 1.0
    for i in range(len(sentence)):
        if i < 1:
            continue
            
        context = sentence[i-1] # słowo poprzedzające
        word = sentence[i]      # aktualne słowo
        
        if context in markov_dict.keys():        # jeśli słowo kontekstowe występuje w modelu - OK
            if word in markov_dict[context].keys(): # jeśli słowo 'word' współwystępowało z 'context' w korpusie - obliczmy prawdopodobieństwo tej sytuacji p(wn|wc)
                prob *= 1.0* markov_dict[context][word]/sum(markov_dict[context].values())
            else:
                prob *= 1/(sum(markov_dict[context].values())+1) # smoothing, jeśli dane slowo 'word' nie występowało po słowie 'context' w korpusie, ustalmy wartość prawdopodobieństwa na bardzo neiwielką. 
    return prob
    

with open('english_europarl.txt', 'r', encoding="utf8") as f:
    text = clean_text(f.read().lower())
    text = make_begin_end_of_sentences(text)
    
    n_grams = get_word_ngrams(text, 2)  # wygeneruj wszystkie 2-gramy słów z korpusu
    for n_gram in n_grams:              # dla każdego n-gramu...
        context = n_gram[-2]            # weź przedostatnie słowo jako kontekst
        last_word = n_gram[-1]          # weź ostatnie słowo jako kontekst
        markov_dict[context].append(last_word) # dopiszmy następniki, które występują w korpusie po kontekście

    for context in markov_dict.keys():                        # dla każdego kontekstu
        markov_dict[context] = Counter(markov_dict[context])  # stwórz histogram słów jakie występują w korpusie po tym kontekście
    
    probability_of_sent1 = get_sentence_probability(text1, markov_dict) # wyznacz prawdopodobieństwo wystąpienia text1
    probability_of_sent2 = get_sentence_probability(text2, markov_dict) # wyznacz prawdopodobieństwo wystąpienia text2
    
    print("Prawdopodobieństwo wystąpienia zdania 1: {p}".format(p=get_sentence_probability(text1, markov_dict)))
    print("Prawdopodobieństwo wystąpienia zdania 2: {p}".format(p=get_sentence_probability(text2, markov_dict)))

Prawdopodobieństwo wystąpienia zdania 1: 5.004813141346049e-20
Prawdopodobieństwo wystąpienia zdania 2: 5.7541016227754555e-24
